src/main/java/reposense/report/RepoCloner.java

```diff
@@ -0,0 +1,117 @@
+        if (isPreviousRepoDifferent()) {
+            isCurrentRepoCloned = spawnCloneProcess(outputPath, config);
+        }
+    }
+
+    /**
+     * Waits for current clone process to finish executing and returns the corresponding {@code RepoConfiguration}.
+     */
+    public RepoConfiguration getClonedRepo(String outputPath) throws IOException {
+        if (isCurrentRepoCloned && isPreviousRepoDifferent()) {
+            isCurrentRepoCloned = waitForCloneProcess(outputPath, configs[index]);
+        }
+        cleanupDirectories();
+
+        if (!isCurrentRepoCloned) {
+            return null;
+        } else if (isPreviousRepoDifferent()) {
+            prevRepoDefaultBranch = GitBranch.getCurrentBranch(configs[index].getRepoRoot());
+        } else {
+            GitClone.updateRepoConfigBranch(configs[index], prevRepoDefaultBranch);
+        }
+        prevIndex = index;
+        index = (index + 1) % configs.length;
+        return configs[prevIndex];
+    }
+
+    /**
+     * Returns true if current repo is different from the previously cloned repo.
+     */
+    private boolean isPreviousRepoDifferent() {
+        return prevIndex == index || !configs[prevIndex].getLocation().equals(configs[index].getLocation());
+    }
+
+    private boolean spawnCloneProcess(String outputPath, RepoConfiguration config) throws IOException {
+        try {
+            GitClone.spawnCloneProcess(config);
+            return true;
+        } catch (GitCloneException gde) {
+            logger.log(Level.WARNING,
+                    "Exception met while trying to clone the repo, will skip this repo.", gde);
+            generateEmptyRepoReport(outputPath, config);
+        } catch (RuntimeException rte) {
+            logger.log(Level.SEVERE, "Error has occurred during analysis, will skip this repo.", rte);
+        }
+        return false;
+    }
+
+    private boolean waitForCloneProcess(String outputPath, RepoConfiguration config) throws IOException {
+        try {
+            GitClone.waitForCloneProcess(config);
+            return true;
+        } catch (GitCloneException gde) {
+            logger.log(Level.WARNING,
+                    "Exception met while trying to clone the repo, will skip this repo.", gde);
+            generateEmptyRepoReport(outputPath, config);
+        } catch (RuntimeException rte) {
+            logger.log(Level.SEVERE, "Error has occurred during analysis, will skip this repo.", rte);
+        }
+        return false;
+    }
+
+    private void generateEmptyRepoReport(String outputPath, RepoConfiguration config) throws IOException {
+        Path repoReportDirectory = Paths.get(outputPath, config.getDisplayName());
+        FileUtil.createDirectory(repoReportDirectory);
+        ReportGenerator.generateEmptyRepoReport(repoReportDirectory.toString());
+    }
+
+    /**
+     * Deletes cloned repo directories that are not in use anymore.
+     */
+    private void cleanupDirectories() {
```

Comment:
why is this called `cleanupDirectories()` when it only cleans up 1 repo?

Reply:
Good point. I originally used plural because it also deletes the folder of repos that failed to clone, but I think thats not a good use of the function.